docs/source/notes/autograd.rst

```diff
@@ -1,124 +1,457 @@
+differentiation will compute the gradients of the loss with respect to the weights of the model.
+It is called the backpropagation algorithm in this case.
+
+
+
+How does it do that?
+^^^^^^^^^^^^^^^^^^^^
+
+One simple interpretation of what it is doing is splitting the user function :math:`f` into composition of smaller
+pre-defined operations.
+We will call such operations elementary.
+The gradient of the original function is then computed by using the chain rule on each of these elementary operations.
+
+For example, if we split :math:`f` into two elementary operations, namely :math:`y = op_2(op_1(x))`,
+the automatic differentiation will compute :math:`v^T J_f = (v^T J_{op_2}) J_{op_1} =  v^T J_{op_2} J_{op_1}`.
+
+What is your code computing?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can request that PyTorch computes gradients for a tensor by passing the keyword argument ``requires_grad=True``
+when constructing the tensor (or, post facto, set this attribute with ``requires_grad_(True)``).
+The user can then use any PyTorch function on this tensor to compute the final value of its function.
+Each of these functions is composed of one or more elementary operation.
+Each operation will add an element to the "chain" if Jacobian products that the backward pass will compute.
+In particular, consider the following function:
+
+.. code::
+
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        return y
+
+    # w is a Tensor of the appropriate size
+    w.requires_grad_()
+
+    # v is a Tensor of the appropriate size
+    f(w).backward(v)
+
+The backward pass will compute the following for a given vector :math:`v`:
+
+.. :math::
+
+    v^T J_f = v^T J_{op_2} J_{op_1} = v^T \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{y}}{\partial{w}}.
+
+Adding a new operation as follows:
+
+.. code::
+
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        z = op_3(y)
+        return z
+
+    w.requires_grad_()
+    f(w).backward(v)
+
+Will change the computation to:
+
+.. math::
+
+    v^T J_f = v^T J_{op_3} J_{op_2} J_{op_1} = v^T \dfrac{\partial{z}}{\partial{y}} \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{z}}{\partial{w}}.
+
+Using a Tensor in multiple places will simply accumulate the gradients from the different branches:
+
+.. code::
+
+    def f(w):
+        x = op_1(w)
+        y = op_2(w)
+        z = op_3(x, y)
+        return z
+
+    w.requires_grad_()
+    f(w).backward(v)
+
+We write the Jacobian of :code:`op_3` by concatenating the Jacobians for each input.
+The computation is then:
+
+.. math::
+
+    v^T J_f = v^T J_{op_3} \begin{pmatrix} J_{op_1} \\ J_{op_2} \end{pmatrix} = v^T \begin{pmatrix} \dfrac{\partial{z}}{\partial{x}} & \dfrac{\partial{z}}{\partial{y}} \end{pmatrix} \begin{pmatrix} \dfrac{\partial{x}}{\partial{w}} \\ \dfrac{\partial{y}}{\partial{w}} \end{pmatrix} = v^T \dfrac{\partial{z}}{\partial{w}}.
+
+Handling in-place operations
```

Comment:
It'll be easier to see what each section address and how we updated the documentation by replacing each subsection title by a question. Thoughts?